Make hardcoded parts of the optimizer sequence configurable (#13376)
ethereum/solidity: docs/yul.rst

Optimization Step Sequence
--------------------------

> By default the optimizer applies its predefined sequence of optimization steps to
> the generated assembly. You can override this sequence and supply your own using
> the ``--yul-optimizations`` option:

Reviewer: Best to break the line after the period:
```suggestion
By default the optimizer applies its predefined sequence of optimization steps to the generated assembly.
You can override this sequence and supply your own using the ``--yul-optimizations`` option:
```

.. code-block:: bash

    solc --optimize --ir-optimized --yul-optimizations 'dhfoD[xarrscLMcCTU]uljmul:fDnTOc'

The sequence inside ``[...]`` will be applied multiple times in a loop until the Yul code
remains unchanged or until the maximum number of rounds (currently 12) has been reached.

> An important thing to note, is that there are hardcoded sequences that are run before and after the

Reviewer: ```suggestion
An important thing to note, is that there are some hardcoded steps that are always run before and after the
```

user-supplied sequence, or the default sequence if one was not supplied by the user.

The cleanup sequence delimiter ``:`` is optional, and is used to supply a custom cleanup sequence
in order to replace the default one. If omitted, the optimizer will simply apply the default cleanup
sequence. In addition, the delimiter may be placed at the beginning of the user-supplied sequence,

> which will be treated as an empty optimization sequence, whereas conversely, if placed at the end of

Reviewer: ```suggestion
which will result in the optimization sequence being empty, whereas conversely, if placed at the end of
```

The cleanup sequence delimiter ``:`` is optional, and is used to supply a custom cleanup sequence
in order to replace the default one. If omitted, the optimizer will simply apply the default cleanup
sequence. In addition, the delimiter may be placed at the beginning of the user-supplied sequence,
which will be treated as an empty optimization sequence, whereas conversely, if placed at the end of
the sequence, will be treated as an empty cleanup sequence.

The order of steps is significant and affects the quality of the output.
Moreover, applying a step may uncover new optimization opportunities for others that were already
applied so repeating steps is often beneficial.
By enclosing part of the sequence in square brackets (``[]``) you tell the optimizer to repeatedly
apply that part until it no longer improves the size of the resulting assembly.

> You can use brackets multiple times in a single sequence but they cannot be nested.

Reviewer: Some of this info is not included in the new text. I guess it was omitted intentionally to simplify the description?

Author: > I guess it was omitted intentionally to simplify the description?

No, I didn't cross check the descriptions in `optimizer.rst` and `yul.rst`, and ended up selecting the `optimizer.rst` description in lieu of the `yul.rst` one. I've added the missing bits.

.. code-block:: sh

>     solc --optimize --ir-optimized --yul-optimizations 'dhfoD[xarrscLMcCTU]uljmul:fDnTOc'

Reviewer 2: I think we should either keep that as before and merely *mention* the delimiter below - or we need to rephrase the paragraph below.

If we keep the delimiter in the example, below should probably rather be something like:

The colon delimiter (``:``) in the example is used to specify a custom cleanup sequence, which is run after the main part of the sequence before the delimiter (and, when using the legacy EVM code transform, only *after* the stack compressor). If no such delimiter is present, the default cleanup sequence (``fDnTOc``) is run after the supplied sequence.

or so...

Author: I'd like to keep the delimiter in the example, since it's now a part of the optimization sequence. I don't really see what the rephrasing actually changes in comparison to the current wording; is it just for style?

Reviewer 2: Without the change the phrasing is contrary to the example...
Then the example *does* have the delimiter and *does not* run any kind of default sequence and only the case that is *not* shown, in which there was *no* delimiter would run a default sequence. That's the difference.
Having the delimiter in the example and still just starting off with the behaviour as if it *wasn't* there, is confusing.
